xds/src/main/java/io/grpc/xds/GcpAuthenticationFilter.java

```diff
@@ -0,0 +1,226 @@
+        // If the CDS resource is not available, fail the RPC with Status.UNAVAILABLE.
+
+        // TODO: Extract the audience from the CDS resource metadata.
+        // If the audience is not found or is in the wrong format, fail the RPC.
+        String audience = "TEST_AUDIENCE";
+
+        try {
+          CallCredentials existingCallCredentials = callOptions.getCredentials();
+          CallCredentials newCallCredentials =
+              getCallCredentials(callCredentialsCache, audience, credentials);
+          if (existingCallCredentials != null) {
+            callOptions = callOptions.withCallCredentials(
+                new CompositeCallCredentials(existingCallCredentials, newCallCredentials));
+          } else {
+            callOptions = callOptions.withCallCredentials(newCallCredentials);
+          }
+        }
+        catch (Exception e) {
+          // If we fail to attach CallCredentials due to any reason, return a FailingClientCall
+          return new FailingClientCall<>(Status.UNAUTHENTICATED
+              .withDescription("Failed to attach CallCredentials.")
+              .withCause(e));
+        }
+        return next.newCall(method, callOptions);
+      }
+    };
+  }
+
+  CallCredentials getCallCredentials(LruCache<String, CallCredentials> cache,
+      String audience, ComputeEngineCredentials credentials) {
+
+    synchronized (cache) {
+      return cache.getOrInsert(audience, key -> {
+        IdTokenCredentials creds = IdTokenCredentials.newBuilder()
+            .setIdTokenProvider(credentials)
+            .setTargetAudience(audience)
+            .build();
+        return MoreCallCredentials.from(creds);
+      });
+    }
+  }
+
+  static final class GcpAuthenticationConfig implements FilterConfig {
+
+    private final long cacheSize;
+
+    public GcpAuthenticationConfig(long cacheSize) {
+      this.cacheSize = cacheSize;
+    }
+
+    public Long getCacheSize() {
+      return cacheSize;
+    }
+
+    @Override
+    public String typeUrl() {
+      return GcpAuthenticationFilter.TYPE_URL;
+    }
+  }
+
+  /** An implementation of {@link ClientCall} that fails when started. */
+  static final class FailingClientCall<ReqT, RespT> extends ClientCall<ReqT, RespT> {
+
+    private final Status error;
+
+    public FailingClientCall(Status error) {
+      this.error = error;
+    }
+
+    @Override
+    public void start(ClientCall.Listener<RespT> listener, Metadata headers) {
+      listener.onClose(error, new Metadata());
+    }
+
+    @Override
+    public void request(int numMessages) {}
+
+    @Override
+    public void cancel(String message, Throwable cause) {}
+
+    @Override
+    public void halfClose() {}
+
+    @Override
+    public void sendMessage(ReqT message) {}
+  }
+
+  static final class LruCache<K, V> {
+
+    private final Map<K, V> cache;
+
+    LruCache(long maxSize) {
+      this.cache = new LinkedHashMap<K, V>(
+          (int) Math.min(maxSize, Integer.MAX_VALUE - 1),
```

Comment:
This needs to be `UnsignedLongs` if you aren't going to limit the values to positive numbers earlier. Really, I suggest doing this min() during config parsing, so there's less code that passes along unsigned longs (and needs to realize that is what it is).